File: macos/src/ffi/Output.swift
import Foundation

enum Output {
    case connectivityChanged(Connectivity)
    case newText(String)
    case error

    static func from(_ output: mpclipboard_Output) -> Self? {
        switch (output.tag) {
        case MPCLIPBOARD_OUTPUT_CONNECTIVITY_CHANGED:
            return .connectivityChanged(Connectivity(output.CONNECTIVITY_CHANGED.connectivity))
        case MPCLIPBOARD_OUTPUT_NEW_TEXT:
            let ptr = output.NEW_TEXT.ptr!
            let len = output.NEW_TEXT.len
            let data = Data(bytes: ptr, count: Int(len))
            if let text = String(data: data, encoding: .utf8) {
                free(ptr)
                return .newText(text)
            } else {
                fatalError("non-utf8 new text in output")
            }
        case MPCLIPBOARD_OUTPUT_IGNORE:
            return nil
        case MPCLIPBOARD_OUTPUT_ERROR:
            return .error
        default:
            fatalError("unnsupported Output")
        }
    }
}

File: android/src/main/kotlin/dev/mpclipboard/android/Output.kt
package dev.mpclipboard.android

sealed interface Output {
    data class ConnectivityChanged(
        val connectivity: Connectivity,
    ) : Output

    data class NewText(
        val text: String,
    ) : Output

    data class Both(
        val connectivity: Connectivity,
        val text: String,
    ) : Output

    companion object {
        internal fun from(nativeOutput: NativeOutput): Output? {
            return when (nativeOutput.tag) {
                Ffi.MPCLIPBOARD_OUTPUT_CONNECTIVITY_CHANGED -> {
                    ConnectivityChanged(Connectivity.from(nativeOutput.connectivity))
                }
                Ffi.MPCLIPBOARD_OUTPUT_NEW_TEXT -> {
                    val text = nativeOutput.text?.toString(Charsets.UTF_8) ?: return null
                    NewText(text)
                }
                Ffi.MPCLIPBOARD_OUTPUT_BOTH -> {
                    val text = nativeOutput.text?.toString(Charsets.UTF_8) ?: return null
                    Both(Connectivity.from(nativeOutput.connectivity), text)
                }
                Ffi.MPCLIPBOARD_OUTPUT_IGNORE -> null
                Ffi.MPCLIPBOARD_OUTPUT_ERROR -> throw RuntimeException("mpclipboard_read failed")
                else -> throw RuntimeException("unknown native output tag: ${nativeOutput.tag}")
            }
        }
    }
}
